MigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withRestart[backups:1,nodes:7]

```
java.lang.AssertionError: Partition: 14 is leaking on Address[127.0.0.1]:9266
Expected: (a value equal to or greater than <0> and a value less than or equal to <1>)
     but: a value less than or equal to <1> <5> was greater than <1>
    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
    at org.junit.Assert.assertThat(Assert.java:956)
    at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$4.run(PartitionCorrectnessTestSupport.java:165)
    at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:857)
    at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:871)
    at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:145)
    at com.hazelcast.internal.partition.MigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withRestart(MigrationCorrectnessTest.java:144)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK7/com.hazelcast$hazelcast/184/testReport/junit/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_withRestart_backups_1_nodes_7_/

Output: 
https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK7/184/artifact/hazelcast/target/surefire-reports/com.hazelcast.internal.partition.MigrationCorrectnessTest-output.txt

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/MigrationInfo.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/MigrationInfo.java
@@ -68,7 +68,9 @@ public class MigrationInfo implements DataSerializable {
     private int partitionId;
 
     private Address source;
+    private String sourceUuid;
     private Address destination;
+    private String destinationUuid;
     private Address master;
 
     private int sourceCurrentReplicaIndex;
@@ -82,12 +84,15 @@ public class MigrationInfo implements DataSerializable {
     public MigrationInfo() {
     }
 
-    public MigrationInfo(int partitionId, Address source, Address destination, int sourceCurrentReplicaIndex,
-                         int sourceNewReplicaIndex, int destinationCurrentReplicaIndex, int destinationNewReplicaIndex) {
+    public MigrationInfo(int partitionId, Address source, String sourceUuid, Address destination, String destinationUuid,
+                         int sourceCurrentReplicaIndex, int sourceNewReplicaIndex,
+                         int destinationCurrentReplicaIndex, int destinationNewReplicaIndex) {
         this.uuid = UuidUtil.newUnsecureUuidString();
         this.partitionId = partitionId;
         this.source = source;
+        this.sourceUuid = sourceUuid;
         this.destination = destination;
+        this.destinationUuid = destinationUuid;
         this.sourceCurrentReplicaIndex = sourceCurrentReplicaIndex;
         this.sourceNewReplicaIndex = sourceNewReplicaIndex;
         this.destinationCurrentReplicaIndex = destinationCurrentReplicaIndex;
@@ -99,8 +104,16 @@ public class MigrationInfo implements DataSerializable {
         return source;
     }
 
+    public String getSourceUuid() {
+        return sourceUuid;
+    }
+
     public Address getDestination() {
         return destination;
+    }
+
+    public String getDestinationUuid() {
+        return destinationUuid;
     }
 
     public int getPartitionId() {
@@ -171,12 +184,14 @@ public class MigrationInfo implements DataSerializable {
         out.writeBoolean(hasSource);
         if (hasSource) {
             source.writeData(out);
+            out.writeUTF(sourceUuid);
         }
 
         boolean hasDestination = destination != null;
         out.writeBoolean(hasDestination);
         if (hasDestination) {
             destination.writeData(out);
+            out.writeUTF(destinationUuid);
         }
 
         master.writeData(out);
@@ -196,12 +211,14 @@ public class MigrationInfo implements DataSerializable {
         if (hasSource) {
             source = new Address();
             source.readData(in);
+            sourceUuid = in.readUTF();
         }
 
         boolean hasDestination = in.readBoolean();
         if (hasDestination) {
             destination = new Address();
             destination.readData(in);
+            destinationUuid = in.readUTF();
         }
 
         master = new Address();
@@ -234,9 +251,11 @@ public class MigrationInfo implements DataSerializable {
         sb.append("uuid=").append(uuid);
         sb.append(", partitionId=").append(partitionId);
         sb.append(", source=").append(source);
+        sb.append(", sourceUuid=").append(sourceUuid);
         sb.append(", sourceCurrentReplicaIndex=").append(sourceCurrentReplicaIndex);
         sb.append(", sourceNewReplicaIndex=").append(sourceNewReplicaIndex);
         sb.append(", destination=").append(destination);
+        sb.append(", destinationUuid=").append(destinationUuid);
         sb.append(", destinationCurrentReplicaIndex=").append(destinationCurrentReplicaIndex);
         sb.append(", destinationNewReplicaIndex=").append(destinationNewReplicaIndex);
         sb.append(", master=").append(master);

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/PartitionRuntimeState.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/PartitionRuntimeState.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.internal.partition;
 
+import com.hazelcast.core.Member;
 import com.hazelcast.internal.cluster.MemberInfo;
 import com.hazelcast.internal.partition.impl.PartitionDataSerializerHook;
 import com.hazelcast.logging.ILogger;
@@ -155,6 +156,16 @@ public final class PartitionRuntimeState implements IdentifiedDataSerializable {
         this.completedMigrations = completedMigrations;
     }
 
+    public boolean isKnownOrNewMember(Member member) {
+        for (MemberInfo m : members) {
+            if (member.getAddress().equals(m.getAddress())) {
+                return member.getUuid().equals(m.getUuid());
+            }
+        }
+
+        return true;
+    }
+
     @Override
     public void readData(ObjectDataInput in) throws IOException {
         version = in.readInt();
@@ -253,4 +264,5 @@ public final class PartitionRuntimeState implements IdentifiedDataSerializable {
     public int getId() {
         return PartitionDataSerializerHook.PARTITION_STATE;
     }
+
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java
@@ -547,6 +547,8 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
             }
         }
 
+
+
         return applyNewState(partitionState, sender);
     }
 
@@ -567,6 +569,12 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
                 }
 
                 return true;
+            }
+
+            if (node.getClusterService().getClusterState() == ClusterState.ACTIVE
+                    && !partitionState.isKnownOrNewMember(nodeEngine.getLocalMember())) {
+                logger.warning("Ignoring partition table update since this member is not a known member in partition table yet");
+                return false;
             }
 
             partitionStateManager.setVersion(newVersion);

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java
@@ -395,21 +395,26 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
         }
     }
 
-    PartitionRuntimeState createMigrationCommitPartitionState(MigrationInfo migrationInfo) {
+    PartitionRuntimeState createMigrationCommitPartitionState(MigrationInfo... migrationInfos) {
         if (!partitionStateManager.isInitialized()) {
             return null;
         }
         lock.lock();
         try {
             MemberInfo[] members = partitionStateManager.getPartitionTableMembers();
-            List<MigrationInfo> completedMigrations = migrationManager.getCompletedMigrations();
+            List<MigrationInfo> completedMigrations = new ArrayList<MigrationInfo>(migrationManager.getCompletedMigrations());
             ILogger logger = node.getLogger(PartitionRuntimeState.class);
 
             InternalPartition[] partitions = partitionStateManager.getPartitionsCopy();
 
-            int partitionId = migrationInfo.getPartitionId();
-            InternalPartitionImpl partition = (InternalPartitionImpl) partitions[partitionId];
-            migrationManager.applyMigration(partition, migrationInfo);
+            for (MigrationInfo migrationInfo : migrationInfos) {
+                int partitionId = migrationInfo.getPartitionId();
+                InternalPartitionImpl partition = (InternalPartitionImpl) partitions[partitionId];
+                migrationManager.applyMigration(partition, migrationInfo);
+
+                migrationInfo.setStatus(MigrationStatus.SUCCESS);
+                completedMigrations.add(migrationInfo);
+            }
 
             int committedVersion = getPartitionStateVersion() + 1;
             return new PartitionRuntimeState(logger, members, partitions, completedMigrations, committedVersion);

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
@@ -17,6 +17,7 @@
 package com.hazelcast.internal.partition.impl;
 
 import com.hazelcast.cluster.ClusterState;
+import com.hazelcast.core.HazelcastInstanceNotActiveException;
 import com.hazelcast.core.MemberLeftException;
 import com.hazelcast.core.MigrationEvent;
 import com.hazelcast.instance.MemberImpl;
@@ -52,11 +53,13 @@ import com.hazelcast.util.scheduler.CoalescingDelayedTrigger;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedHashSet;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Map;
 import java.util.Queue;
 import java.util.Set;
 import java.util.concurrent.ExecutionException;
@@ -301,9 +304,10 @@ public class MigrationManager {
                     && node.getThisAddress().equals(migrationInfo.getDestination())
                     && migrationInfo.getDestinationCurrentReplicaIndex() > 0
                     && migrationInfo.getDestinationNewReplicaIndex() == 0) {
+
                 // shift up when a partition owner is removed
                 assert migrationInfo.getStatus() == MigrationStatus.SUCCESS
-                        : "Promotion status should be SUCCESS! -> " + migrationInfo.getStatus();
+                        : "Promotion status should be SUCCESS! -> " + migrationInfo;
                 PromoteFromBackupOperation op = new PromoteFromBackupOperation(migrationInfo.getDestinationCurrentReplicaIndex());
                 op.setPartitionId(migrationInfo.getPartitionId()).setNodeEngine(nodeEngine).setService(partitionService);
                 nodeEngine.getOperationService().executeOperation(op);
@@ -314,45 +318,53 @@ public class MigrationManager {
         }
     }
 
-    private boolean commitMigrationToDestination(MigrationInfo migrationInfo) {
+    private boolean commitMigrationToDestination(Address destination, MigrationInfo... migrations) {
+        assert migrations.length > 0 : "No migrations to commit!";
         if (!node.isMaster()) {
             return false;
         }
 
-        if (node.getThisAddress().equals(migrationInfo.getDestination())) {
-            final boolean result = node.isRunning();
+        if (node.getThisAddress().equals(destination)) {
             if (logger.isFinestEnabled()) {
-                logger.finest("Shortcutting migration commit with result: " + result
-                        + " since destination is master: " + migrationInfo);
+                logger.finest("Shortcutting migration commit,"
+                        + " since destination is master. Migrations " + Arrays.toString(migrations));
             }
-            return result;
+            return true;
         }
 
         try {
             if (logger.isFinestEnabled()) {
-                logger.finest("Sending commit operation for: " + migrationInfo);
+                logger.finest("Sending commit operation to " + destination + " for " + Arrays.toString(migrations));
             }
-            PartitionRuntimeState partitionState = partitionService.createMigrationCommitPartitionState(migrationInfo);
-            MigrationCommitOperation operation = new MigrationCommitOperation(partitionState, migrationInfo);
+            PartitionRuntimeState partitionState = partitionService.createMigrationCommitPartitionState(migrations);
+            String destinationUuid = partitionStateManager.getMemberUuid(destination);
+            MigrationCommitOperation operation = new MigrationCommitOperation(partitionState, destinationUuid);
             Future<Boolean> future = nodeEngine.getOperationService()
-                    .createInvocationBuilder(SERVICE_NAME, operation,
-                            migrationInfo.getDestination())
+                    .createInvocationBuilder(SERVICE_NAME, operation, destination)
                     .setTryCount(Integer.MAX_VALUE)
                     .setCallTimeout(Long.MAX_VALUE).invoke();
-            final Boolean result = future.get();
+
+            boolean result = future.get();
             if (logger.isFinestEnabled()) {
-                logger.finest("Migration commit result: " + result + " for migration : " + migrationInfo);
+                logger.finest("Migration commit result " + result + " from " + destination
+                        + " for migrations " + Arrays.toString(migrations));
             }
             return result;
         } catch (Throwable t) {
-            if (t instanceof MemberLeftException || t instanceof TargetNotMemberException) {
-                logger.warning("Migration commit failed for " + migrationInfo + " since destination left the cluster");
-            } else {
-                logger.severe("Migration commit failed for " + migrationInfo, t);
-            }
+            boolean memberLeft = t instanceof MemberLeftException
+                        || t.getCause() instanceof TargetNotMemberException
+                        || t.getCause() instanceof HazelcastInstanceNotActiveException;
 
-            return false;
+            if (memberLeft) {
+                logger.warning("Migration commit failed for "
+                        + (migrations.length == 1 ? migrations[0] : "multiple migrations")
+                        + " since destination " + destination + " left the cluster");
+            } else {
+                logger.severe("Migration commit to " + destination + " failed for "
+                        + (migrations.length == 1 ? migrations[0] : "multiple migrations"), t);
+            }
         }
+        return false;
     }
 
     boolean addCompletedMigration(MigrationInfo migrationInfo) {
@@ -492,51 +504,6 @@ public class MigrationManager {
 
     void scheduleMigration(MigrationInfo migrationInfo) {
         migrationQueue.add(new MigrateTask(migrationInfo));
-    }
-
-    private void repairPartition(int partitionId) {
-        InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(partitionId);
-
-        if (partition.getOwnerOrNull() == null) {
-            Address destination = null;
-            int index = 1;
-            for (int i = index; i < InternalPartition.MAX_REPLICA_COUNT; i++) {
-                destination = partition.getReplicaAddress(i);
-                if (destination != null) {
-                    partition.swapAddresses(0, i);
-                    index = i;
-                    break;
-                }
-            }
-
-            if (logger.isFinestEnabled()) {
-                if (destination != null) {
-                    logger.finest("partitionId=" + partition.getPartitionId() + " owner is removed. replicaIndex=" + index
-                            + " is shifted up to 0. " + partition);
-                } else {
-                    logger.finest("partitionId=" + partition.getPartitionId()
-                            + " owner is removed. there is no other replica to shift up. " + partition);
-                }
-            }
-
-            if (destination != null) {
-                // Promotion will be executed when new partition owner gets this completed migration.
-                // TODO destination uuid?
-                String destinationeUuid = partitionStateManager.getMemberUuid(destination);
-                MigrationInfo migration =
-                        new MigrationInfo(partitionId, null, null, destination, destinationeUuid, -1, -1, index, 0);
-                migration.setMaster(node.getThisAddress());
-                migration.setStatus(MigrationInfo.MigrationStatus.SUCCESS);
-                partitionService.getMigrationManager().addCompletedMigration(migration);
-                partitionService.getMigrationManager().scheduleActiveMigrationFinalization(migration);
-            }
-        }
-
-        if (partition.getOwnerOrNull() == null) {
-            logger.warning("partitionId=" + partitionId + " is completely lost!");
-            PartitionEventManager partitionEventManager = partitionService.getPartitionEventManager();
-            partitionEventManager.sendPartitionLostEvent(partitionId, InternalPartition.MAX_BACKUP_COUNT);
-        }
     }
 
     void applyMigration(InternalPartitionImpl partition, MigrationInfo migrationInfo) {
@@ -814,20 +781,21 @@ public class MigrationManager {
                 return;
             }
 
+            if (migrationInfo.getSource() == null
+                    && migrationInfo.getDestinationCurrentReplicaIndex() > 0
+                    && migrationInfo.getDestinationNewReplicaIndex() == 0) {
+
+                throw new AssertionError("Promotion migrations should be handled by "
+                        + RepairPartitionTableTask.class.getSimpleName() + "! -> " + migrationInfo);
+            }
+
             try {
                 MemberImpl partitionOwner = checkMigrationParticipantsAndGetPartitionOwner();
                 if (partitionOwner == null) {
                     return;
                 }
 
-                internalMigrationListener.onMigrationStart(MigrationParticipant.MASTER, migrationInfo);
-                partitionService.getPartitionEventManager()
-                        .sendMigrationEvent(migrationInfo, MigrationEvent.MigrationStatus.STARTED);
-
-                if (logger.isFineEnabled()) {
-                    logger.fine("Starting Migration: " + migrationInfo);
-                }
-
+                beforeMigration();
                 Boolean result = executeMigrateOperation(partitionOwner);
                 processMigrationResult(result);
             } catch (Throwable t) {
@@ -835,6 +803,16 @@ public class MigrationManager {
                 logger.log(level, "Error [" + t.getClass() + ": " + t.getMessage() + "] during " + migrationInfo);
                 logger.finest(t);
                 migrationOperationFailed();
+            }
+        }
+
+        private void beforeMigration() {
+            internalMigrationListener.onMigrationStart(MigrationParticipant.MASTER, migrationInfo);
+            partitionService.getPartitionEventManager()
+                    .sendMigrationEvent(migrationInfo, MigrationEvent.MigrationStatus.STARTED);
+
+            if (logger.isFineEnabled()) {
+                logger.fine("Starting Migration: " + migrationInfo);
             }
         }
 
@@ -959,15 +937,14 @@ public class MigrationManager {
         }
 
         private void migrationOperationSucceeded() {
-            migrationInfo.setStatus(MigrationStatus.SUCCESS);
             internalMigrationListener.onMigrationComplete(MigrationParticipant.MASTER, migrationInfo, true);
-            addCompletedMigration(migrationInfo);
 
-            boolean commitSuccessful = commitMigrationToDestination(migrationInfo);
+            boolean commitSuccessful = commitMigrationToDestination(migrationInfo.getDestination(), migrationInfo);
 
             partitionServiceLock.lock();
             try {
                 if (commitSuccessful) {
+                    migrationInfo.setStatus(MigrationStatus.SUCCESS);
                     internalMigrationListener.onMigrationCommit(MigrationParticipant.MASTER, migrationInfo);
 
                     // updates partition table after successful commit
@@ -982,6 +959,7 @@ public class MigrationManager {
                     triggerRepartitioningAfterMigrationFailure();
                 }
 
+                addCompletedMigration(migrationInfo);
                 scheduleActiveMigrationFinalization(migrationInfo);
                 if (partitionService.syncPartitionRuntimeState()) {
                     evictCompletedMigrations(migrationInfo);
@@ -1010,26 +988,130 @@ public class MigrationManager {
             partitionServiceLock.lock();
             try {
                 Collection<Address> addresses = collectUnknownAddresses();
-
                 for (Address address : addresses) {
                     partitionStateManager.removeDeadAddress(address);
                 }
 
-                for (int partitionId = 0; partitionId < partitionService.getPartitionCount(); partitionId++) {
-                    repairPartition(partitionId);
+                if (!promoteBackupsForMissingOwners()) {
+                    triggerControlTask();
+                    return;
                 }
-
-                partitionService.syncPartitionRuntimeState();
 
                 if (logger.isFinestEnabled()) {
                     logger.finest("RepartitioningTask scheduled");
                 }
-
                 migrationQueue.add(new RepartitioningTask());
 
             } finally {
                 partitionServiceLock.unlock();
             }
+        }
+
+        private boolean promoteBackupsForMissingOwners() {
+            Map<Address, Collection<MigrationInfo>> promotions = new HashMap<Address, Collection<MigrationInfo>>();
+            for (int partitionId = 0; partitionId < partitionService.getPartitionCount(); partitionId++) {
+                MigrationInfo migration = createPromotionMigrationIfOwnerIsNull(partitionId);
+                if (migration == null) {
+                    continue;
+                }
+
+                Collection<MigrationInfo> migrations = promotions.get(migration.getDestination());
+                if (migrations == null) {
+                    migrations = new ArrayList<MigrationInfo>();
+                    promotions.put(migration.getDestination(), migrations);
+                }
+                migrations.add(migration);
+            }
+
+            boolean allSucceeded = true;
+            for (Map.Entry<Address, Collection<MigrationInfo>> entry : promotions.entrySet()) {
+                Address destination = entry.getKey();
+                Collection<MigrationInfo> migrations = entry.getValue();
+
+                allSucceeded &= commitPromotionMigrations(destination, migrations);
+            }
+            return allSucceeded;
+        }
+
+        private boolean commitPromotionMigrations(Address destination, Collection<MigrationInfo> migrations) {
+            boolean success;
+            if (node.getThisAddress().equals(destination)) {
+                if (logger.isFinestEnabled()) {
+                    logger.finest("Shortcutting promotions with result: true, "
+                            + "since destination is master.");
+                }
+                success = true;
+            } else {
+                MigrationInfo[] migrationsArray = migrations.toArray(new MigrationInfo[migrations.size()]);
+                success = commitMigrationToDestination(destination, migrationsArray);
+            }
+
+            for (MigrationInfo migration : migrations) {
+                if (success) {
+                    InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
+
+                    assert partition.getOwnerOrNull() == null : "Owner should be null: " + partition;
+                    assert destination.equals(partition
+                            .getReplicaAddress(migration.getDestinationCurrentReplicaIndex()))
+                            : "Invalid replica! Destination: " + destination
+                                + ", index: " + migration.getDestinationCurrentReplicaIndex() + ", " + partition;
+
+                    partition.swapAddresses(0, migration.getDestinationCurrentReplicaIndex());
+                    addCompletedMigration(migration);
+                    scheduleActiveMigrationFinalization(migration);
+                }
+            }
+
+            if (!success) {
+                int delta = PARTITION_STATE_VERSION_INCREMENT_DELTA_ON_MIGRATION_FAILURE;
+                partitionService.getPartitionStateManager().incrementVersion(delta);
+            }
+
+            partitionService.syncPartitionRuntimeState();
+            return success;
+        }
+
+        private MigrationInfo createPromotionMigrationIfOwnerIsNull(int partitionId) {
+            InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(partitionId);
+
+            if (partition.getOwnerOrNull() == null) {
+                Address destination = null;
+                int index = 1;
+                for (int i = index; i < InternalPartition.MAX_REPLICA_COUNT; i++) {
+                    destination = partition.getReplicaAddress(i);
+                    if (destination != null) {
+                        index = i;
+                        break;
+                    }
+                }
+
+                if (logger.isFinestEnabled()) {
+                    if (destination != null) {
+                        logger.finest("partitionId=" + partition.getPartitionId() + " owner is removed. replicaIndex=" + index
+                                + " will be shifted up to 0. " + partition);
+                    } else {
+                        logger.finest("partitionId=" + partition.getPartitionId()
+                                + " owner is removed. there is no other replica to shift up. " + partition);
+                    }
+                }
+
+                if (destination != null) {
+                    String destinationUuid = partitionStateManager.getMemberUuid(destination);
+                    MigrationInfo migration =
+                            new MigrationInfo(partitionId, null, null, destination, destinationUuid, -1, -1, index, 0);
+                    migration.setMaster(node.getThisAddress());
+                    migration.setStatus(MigrationInfo.MigrationStatus.SUCCESS);
+
+                    return migration;
+                }
+            }
+
+            if (partition.getOwnerOrNull() == null) {
+                logger.warning("partitionId=" + partitionId + " is completely lost!");
+                PartitionEventManager partitionEventManager = partitionService.getPartitionEventManager();
+                partitionEventManager.sendPartitionLostEvent(partitionId, InternalPartition.MAX_BACKUP_COUNT);
+            }
+            return null;
         }
 
         private Collection<Address> collectUnknownAddresses() {
@@ -1054,7 +1136,6 @@ public class MigrationManager {
 
             return addresses;
         }
-
     }
 
     private class ControlTask implements MigrationRunnable {

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
@@ -333,7 +333,7 @@ public class MigrationManager {
                 logger.finest("Sending commit operation for: " + migrationInfo);
             }
             PartitionRuntimeState partitionState = partitionService.createMigrationCommitPartitionState(migrationInfo);
-            MigrationCommitOperation operation = new MigrationCommitOperation(partitionState);
+            MigrationCommitOperation operation = new MigrationCommitOperation(partitionState, migrationInfo);
             Future<Boolean> future = nodeEngine.getOperationService()
                     .createInvocationBuilder(SERVICE_NAME, operation,
                             migrationInfo.getDestination())
@@ -343,7 +343,7 @@ public class MigrationManager {
             if (logger.isFinestEnabled()) {
                 logger.finest("Migration commit result: " + result + " for migration : " + migrationInfo);
             }
-            return true;
+            return result;
         } catch (Throwable t) {
             if (t instanceof MemberLeftException || t instanceof TargetNotMemberException) {
                 logger.warning("Migration commit failed for " + migrationInfo + " since destination left the cluster");
@@ -521,8 +521,10 @@ public class MigrationManager {
 
             if (destination != null) {
                 // Promotion will be executed when new partition owner gets this completed migration.
+                // TODO destination uuid?
+                String destinationeUuid = partitionStateManager.getMemberUuid(destination);
                 MigrationInfo migration =
-                        new MigrationInfo(partition.getPartitionId(), null, destination, -1, -1, index, 0);
+                        new MigrationInfo(partitionId, null, null, destination, destinationeUuid, -1, -1, index, 0);
                 migration.setMaster(node.getThisAddress());
                 migration.setStatus(MigrationInfo.MigrationStatus.SUCCESS);
                 partitionService.getMigrationManager().addCompletedMigration(migration);
@@ -673,7 +675,8 @@ public class MigrationManager {
         }
 
         private void assignNewPartitionOwner(int partitionId, InternalPartitionImpl currentPartition, Address newOwner) {
-            MigrationInfo migrationInfo = new MigrationInfo(partitionId, null, newOwner, -1, -1, -1, 0);
+            String destinationUuid = partitionStateManager.getMemberUuid(newOwner);
+            MigrationInfo migrationInfo = new MigrationInfo(partitionId, null, null, newOwner, destinationUuid, -1, -1, -1, 0);
             PartitionEventManager partitionEventManager = partitionService.getPartitionEventManager();
             partitionEventManager.sendMigrationEvent(migrationInfo, MigrationEvent.MigrationStatus.STARTED);
             currentPartition.setReplicaAddress(0, newOwner);
@@ -741,8 +744,11 @@ public class MigrationManager {
 
                     partition.setReplicaAddress(sourceCurrentReplicaIndex, null);
                 } else {
-                    MigrationInfo migration = new MigrationInfo(partitionId, source, destination, sourceCurrentReplicaIndex,
-                            sourceNewReplicaIndex, destinationCurrentReplicaIndex, destinationNewReplicaIndex);
+                    String sourceUuid = partitionStateManager.getMemberUuid(source);
+                    String destinationUuid = partitionStateManager.getMemberUuid(destination);
+                    MigrationInfo migration = new MigrationInfo(partitionId, source, sourceUuid, destination, destinationUuid,
+                            sourceCurrentReplicaIndex, sourceNewReplicaIndex,
+                            destinationCurrentReplicaIndex, destinationNewReplicaIndex);
 
                     migrationCount.value++;
                     migrations.offer(migration);

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java
@@ -196,6 +196,10 @@ public class PartitionStateManager {
         }
     }
 
+    String getMemberUuid(Address address) {
+        return address != null ? memberUuidMap.get(address) : null;
+    }
+
     boolean isKnownMemberUuid(Address address, String uuid) {
         String currentUuid = memberUuidMap.get(address);
         assert currentUuid != null : "Unknown Member! " + address + " - " + uuid;

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BaseMigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BaseMigrationOperation.java
@@ -17,6 +17,7 @@
 package com.hazelcast.internal.partition.operation;
 
 import com.hazelcast.cluster.ClusterState;
+import com.hazelcast.core.Member;
 import com.hazelcast.core.MemberLeftException;
 import com.hazelcast.instance.Node;
 import com.hazelcast.internal.partition.InternalPartition;
@@ -58,6 +59,7 @@ abstract class BaseMigrationOperation extends AbstractOperation
         try {
             onMigrationStart();
             verifyPartitionStateVersion();
+            verifyMemberUuid();
             verifyClusterState();
         } catch (Exception e) {
             onMigrationComplete(false);
@@ -75,6 +77,21 @@ abstract class BaseMigrationOperation extends AbstractOperation
 
             // this is expected when cluster member list changes during migration
             throw new PartitionStateVersionMismatchException(partitionStateVersion, localPartitionStateVersion);
+        }
+    }
+
+    private void verifyMemberUuid() {
+        Member localMember = getNodeEngine().getLocalMember();
+        if (localMember.getAddress().equals(migrationInfo.getSource())) {
+            if (!localMember.getUuid().equals(migrationInfo.getSourceUuid())) {
+                throw new IllegalStateException("This member " + localMember
+                        + " is the migration source but has a different uuid! Migration: " + migrationInfo);
+            }
+        } else if (localMember.getAddress().equals(migrationInfo.getDestination())) {
+            if (!localMember.getUuid().equals(migrationInfo.getDestinationUuid())) {
+                throw new IllegalStateException("This member " + localMember
+                        + " is the migration destination but has a different uuid! Migration: " + migrationInfo);
+            }
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationCommitOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationCommitOperation.java
@@ -17,9 +17,11 @@
 package com.hazelcast.internal.partition.operation;
 
 import com.hazelcast.core.HazelcastInstanceNotActiveException;
+import com.hazelcast.core.Member;
 import com.hazelcast.core.MemberLeftException;
 import com.hazelcast.internal.partition.InternalPartitionService;
 import com.hazelcast.internal.partition.MigrationCycleOperation;
+import com.hazelcast.internal.partition.MigrationInfo;
 import com.hazelcast.internal.partition.PartitionRuntimeState;
 import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
 import com.hazelcast.nio.ObjectDataInput;
@@ -41,11 +43,16 @@ public class MigrationCommitOperation extends AbstractOperation implements Migra
 
     private PartitionRuntimeState partitionState;
 
+    private MigrationInfo migrationInfo;
+
+    private boolean success;
+
     public MigrationCommitOperation() {
     }
 
-    public MigrationCommitOperation(PartitionRuntimeState partitionState) {
+    public MigrationCommitOperation(PartitionRuntimeState partitionState, MigrationInfo migrationInfo) {
         this.partitionState = partitionState;
+        this.migrationInfo = migrationInfo;
     }
 
     @Override
@@ -53,11 +60,17 @@ public class MigrationCommitOperation extends AbstractOperation implements Migra
         NodeEngine nodeEngine = getNodeEngine();
         if (!nodeEngine.isRunning()) {
             throw new HazelcastInstanceNotActiveException("This node is shutting down!");
+        } else {
+            final Member localMember = nodeEngine.getLocalMember();
+            if (!localMember.getUuid().equals(migrationInfo.getDestinationUuid())) {
+                throw new IllegalStateException("This member " + localMember
+                        + " is migration destination but it is restarted. Migration: " + migrationInfo);
+            }
         }
 
         partitionState.setEndpoint(getCallerAddress());
         InternalPartitionServiceImpl partitionService = getService();
-        partitionService.processPartitionRuntimeState(partitionState);
+        success = partitionService.processPartitionRuntimeState(partitionState);
     }
 
     @Override
@@ -67,7 +80,7 @@ public class MigrationCommitOperation extends AbstractOperation implements Migra
 
     @Override
     public Object getResponse() {
-        return Boolean.TRUE;
+        return success;
     }
 
     @Override
@@ -89,11 +102,14 @@ public class MigrationCommitOperation extends AbstractOperation implements Migra
         super.readInternal(in);
         partitionState = new PartitionRuntimeState();
         partitionState.readData(in);
+        migrationInfo = new MigrationInfo();
+        migrationInfo.readData(in);
     }
 
     @Override
     protected void writeInternal(ObjectDataOutput out) throws IOException {
         super.writeInternal(out);
         partitionState.writeData(out);
+        migrationInfo.writeData(out);
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationCommitOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationCommitOperation.java
@@ -16,12 +16,10 @@
 
 package com.hazelcast.internal.partition.operation;
 
-import com.hazelcast.core.HazelcastInstanceNotActiveException;
 import com.hazelcast.core.Member;
 import com.hazelcast.core.MemberLeftException;
 import com.hazelcast.internal.partition.InternalPartitionService;
 import com.hazelcast.internal.partition.MigrationCycleOperation;
-import com.hazelcast.internal.partition.MigrationInfo;
 import com.hazelcast.internal.partition.PartitionRuntimeState;
 import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
 import com.hazelcast.nio.ObjectDataInput;
@@ -43,29 +41,26 @@ public class MigrationCommitOperation extends AbstractOperation implements Migra
 
     private PartitionRuntimeState partitionState;
 
-    private MigrationInfo migrationInfo;
+    private String expectedMemberUuid;
 
     private boolean success;
 
     public MigrationCommitOperation() {
     }
 
-    public MigrationCommitOperation(PartitionRuntimeState partitionState, MigrationInfo migrationInfo) {
+    public MigrationCommitOperation(PartitionRuntimeState partitionState, String expectedMemberUuid) {
         this.partitionState = partitionState;
-        this.migrationInfo = migrationInfo;
+        this.expectedMemberUuid = expectedMemberUuid;
     }
 
     @Override
     public void run() {
         NodeEngine nodeEngine = getNodeEngine();
-        if (!nodeEngine.isRunning()) {
-            throw new HazelcastInstanceNotActiveException("This node is shutting down!");
-        } else {
-            final Member localMember = nodeEngine.getLocalMember();
-            if (!localMember.getUuid().equals(migrationInfo.getDestinationUuid())) {
-                throw new IllegalStateException("This member " + localMember
-                        + " is migration destination but it is restarted. Migration: " + migrationInfo);
-            }
+        final Member localMember = nodeEngine.getLocalMember();
+        if (!localMember.getUuid().equals(expectedMemberUuid)) {
+            throw new IllegalStateException("This " + localMember
+                    + " is migration commit destination but most probably it's restarted "
+                    + "and not the expected target.");
         }
 
         partitionState.setEndpoint(getCallerAddress());
@@ -100,16 +95,15 @@ public class MigrationCommitOperation extends AbstractOperation implements Migra
     @Override
     protected void readInternal(ObjectDataInput in) throws IOException {
         super.readInternal(in);
+        expectedMemberUuid = in.readUTF();
         partitionState = new PartitionRuntimeState();
         partitionState.readData(in);
-        migrationInfo = new MigrationInfo();
-        migrationInfo.readData(in);
     }
 
     @Override
     protected void writeInternal(ObjectDataOutput out) throws IOException {
         super.writeInternal(out);
+        out.writeUTF(expectedMemberUuid);
         partitionState.writeData(out);
-        migrationInfo.writeData(out);
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/MigrationCommitServiceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/MigrationCommitServiceTest.java
@@ -301,26 +301,38 @@ public class MigrationCommitServiceTest
     private MigrationInfo createMoveMigration(final int partitionId, final int replicaIndex, final Address destination) {
         final InternalPartition partition = getPartition(instances[0], partitionId);
         final Address source = partition.getReplicaAddress(replicaIndex);
-        return new MigrationInfo(partitionId, source, destination, replicaIndex, -1, -1, replicaIndex);
+        final String sourceUuid = getMemberUuid(source);
+        final String destinationUuid = getMemberUuid(destination);
+
+        return new MigrationInfo(partitionId, source, sourceUuid, destination, destinationUuid, replicaIndex, -1, -1, replicaIndex);
     }
 
     private MigrationInfo createShiftDownMigration(final int partitionId, final int oldReplicaIndex, final int newReplicaIndex,
                                                    final Address destination) {
         final InternalPartitionImpl partition = getPartition(instances[0], partitionId);
         final Address source = partition.getReplicaAddress(oldReplicaIndex);
-        return new MigrationInfo(partitionId, source, destination, oldReplicaIndex, newReplicaIndex, -1,
+        final String sourceUuid = getMemberUuid(source);
+        final String destinationUuid = getMemberUuid(destination);
+
+        return new MigrationInfo(partitionId, source, sourceUuid, destination, destinationUuid, oldReplicaIndex, newReplicaIndex, -1,
                 oldReplicaIndex);
     }
 
     private MigrationInfo createCopyMigration(final int partitionId, final int copyReplicaIndex, final Address destination) {
-        return new MigrationInfo(partitionId, null, destination, -1, -1, -1, copyReplicaIndex);
+        return new MigrationInfo(partitionId, null, null, destination, getMemberUuid(destination), -1, -1, -1, copyReplicaIndex);
     }
 
     private MigrationInfo createShiftUpMigration(final int partitionId, final int oldReplicaIndex, final int newReplicaIndex) {
         final InternalPartitionImpl partition = getPartition(instances[0], partitionId);
         final Address source = partition.getReplicaAddress(newReplicaIndex);
+        final String sourceUuid = getMemberUuid(source);
         final Address destination = partition.getReplicaAddress(oldReplicaIndex);
-        return new MigrationInfo(partitionId, source, destination, newReplicaIndex, -1, oldReplicaIndex, newReplicaIndex);
+        final String destinationUuid = getMemberUuid(destination);
+        return new MigrationInfo(partitionId, source, sourceUuid, destination, destinationUuid, newReplicaIndex, -1, oldReplicaIndex, newReplicaIndex);
+    }
+
+    private String getMemberUuid(Address address) {
+        return address != null ? getNodeEngineImpl(factory.getInstance(address)).getLocalMember().getUuid() : null;
     }
 
     private Address clearReplicaIndex(final int partitionId, final int replicaIndexToClear)

